src/shell/xdg/window/mod.rs

```diff
@@ -9,28 +9,124 @@ use wayland_client::{
+    zxdg_decoration_manager_v1: Option<(u32, zxdg_decoration_manager_v1::ZxdgDecorationManagerV1)>,
+    windows: Vec<Window>,
+}
+
+impl XdgWindowState {
+    pub fn new() -> XdgWindowState {
+        XdgWindowState { xdg_wm_base: None, zxdg_decoration_manager_v1: None, windows: vec![] }
+    }
+
+    pub fn window_by_wl(&self, surface: &wl_surface::WlSurface) -> Option<&Window> {
+        self.windows.iter().find(|window| window.wl_surface() == surface)
+    }
+
+    pub fn window_by_xdg(&self, surface: &xdg_surface::XdgSurface) -> Option<&Window> {
+        self.windows.iter().find(|window| window.xdg_surface() == surface)
+    }
+
+    pub fn window_by_toplevel(&self, toplevel: &xdg_toplevel::XdgToplevel) -> Option<&Window> {
+        self.windows.iter().find(|window| window.xdg_toplevel() == toplevel)
+    }
+
+    /// Create a window.
+    ///
+    /// This function will create a window from a [`WlSurface`]. Note the window will consume the
+    /// [`WlSurface`] when the window is dropped.
+    ///
+    /// ## Default settings
+    ///
+    /// ### Window decorations
+    ///
+    /// The window will use the decoration mode dictated by the compositor. If you do not want this, set the
+    /// preferred decoration mode before [`map()`](Window::map)ing the window.
+    ///
+    /// ### Initial window size
+    ///
+    /// This will vary depending on the compositor. Some compositors may allow the you to make the window
+    /// any desired size while others may give the you a desired size during the initial commit.
+    ///
+    /// You ultimately have control over what size buffer is committed, meaning you could ignore the
+    /// compositor. However, not respecting the compositor will likely result in aggravated users and a subpar
+    /// experience.
+    ///
+    /// Some compositors may take the minimum and maximum window size in consideration when determining how
+    /// large of a window that will be requested during the initial commit.
+    ///
+    /// [`WlSurface`]: wl_surface::WlSurface
+    #[must_use = "dropping the window will consume the wl_surface and destroy the window"]
+    pub fn create_window<D>(
+        &mut self,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+        wl_surface: wl_surface::WlSurface,
+    ) -> Result<Window, CreateWindowError>
+    where
+        D: Dispatch<wl_surface::WlSurface, UserData = SurfaceData>
+            + Dispatch<xdg_surface::XdgSurface, UserData = ()>
+            + Dispatch<xdg_toplevel::XdgToplevel, UserData = WindowData>
+            + 'static,
+    {
+        // We don't know if an un-managed surface has a role object.
+        let surface_data = wl_surface.data::<SurfaceData>().ok_or(CreateWindowError::HasRole)?;
+
+        // XDG Shell protocol forbids creating an window from a surface that already has a role object.
+        if surface_data.has_role.load(Ordering::SeqCst) {
+            return Err(CreateWindowError::HasRole);
+        }
```

Comment:
I'm not quite sure about those checks, they seem unnecessary as the compositor will check this anyway and trigger a protocol error if necessary, and it seems to me that this kind of condition is *always* a logic bug, so having the function return and error is kind of pointless is the user is always going to `.unwrap()` it anyway. Crashing on the protocol error achieves the same result.

This applies to the whole "SCTK tracking the surface roles itself" actually, not just that particular instance.

Reply:
Considering I've changed the logic to consume the surface, it might not be too nessecary to check this.